src/query/ts/m3db/encoded_block_iterator_builder.go

```diff
@@ -58,31 +70,143 @@ func (b *encodedBlockBuilder) add(
+	if err := b.backfillMissing(); err != nil {
+		return nil, err
+	}
+
 	blocks := make([]block.Block, 0, len(b.blocksAtTime))
-	for _, block := range b.blocksAtTime {
-		block.generateMetas()
+	for _, bl := range b.blocksAtTime {
+		block := bl.block
+		if err := block.generateMetas(); err != nil {
+			return nil, err
+		}
+
 		blocks = append(blocks, &block)
 	}
 
-	return blocks
+	return blocks, nil
+}
+
+// Since some series can be missing if there are no values in composing blocks,
+// need to backfill these, since downstream functions rely on series order.
+//
+// TODO: this will be removed after https://github.com/m3db/m3/issues/1281 is
+// completed, which will allow temporal functions, and final read accumulator
+// to empty series, and do its own matching rather than relying on matching
+// series order. This is functionally throwaway code and should be regarded
+// as such.
+func (b *encodedBlockBuilder) backfillMissing() error {
+	// map series to indeces of b.blocksAtTime at which they are seen
+	seenMap := make(map[string]seriesIteratorDetails, initBlockReplicaLength)
+	for idx, bl := range b.blocksAtTime {
+		block := bl.block
+		for _, iter := range block.seriesBlockIterators {
+			id := iter.ID().String()
+			if seen, found := seenMap[id]; !found {
+				seenMap[id] = seriesIteratorDetails{
+					start:   iter.Start(),
+					end:     iter.End(),
+					id:      iter.ID(),
+					ns:      iter.Namespace(),
+					tagIter: iter.Tags(),
+					present: []int{idx},
+				}
+			} else {
+				seen.present = append(seen.present, idx)
+				seenMap[id] = seen
+			}
+		}
+	}
+
+	// make sure that each seen series exists in every block.
+	blockLen := len(b.blocksAtTime)
+	for _, iterDetails := range seenMap {
+		present := iterDetails.present
+		if len(present) == blockLen {
+			// This series exists in every block already, thus no backfilling necessary.
+			continue
+		}
+
+		for blockIdx, bl := range b.blocksAtTime {
+			found := false
+			for _, presentVal := range present {
+				if presentVal == blockIdx {
+					found = true
+					break
+				}
+			}
+
+			// this series exists in the current block.
+			if found {
+				continue
+			}
+
+			// blockIdx does not contain the present value; need to populate it with
+			// an empty series iterator.
+			tags, err := cloneTagIterator(iterDetails.tagIter)
+			if err != nil {
+				return err
+			}
+
+			iter := encoding.NewSeriesIterator(encoding.SeriesIteratorOptions{
+				ID:             ident.StringID(iterDetails.id.String()),
+				Namespace:      ident.StringID(iterDetails.ns.String()),
+				Tags:           tags,
+				StartInclusive: iterDetails.start,
+				EndExclusive:   iterDetails.end,
+			}, nil)
+
+			block := bl.block
+			block.seriesBlockIterators = append(block.seriesBlockIterators, iter)
+			b.blocksAtTime[blockIdx].block = block
+		}
+	}
+
+	for _, bl := range b.blocksAtTime {
+		sort.Sort(seriesIteratorByID(bl.block.seriesBlockIterators))
```

Comment:
move this out ? I assume you have to do this independent of backfill ?

Reply:
Sounds good. Was originally intending to remove the rest of the backfill logic and just do the sort here, but makes more sense separately.